meetings/working-groups/collection-literals/meetings/working-groups/collection-literals/CL-2023-06-16.md

```diff
@@ -0,0 +1,125 @@
+
+Users who want to be able to return those spans can do so by explicit conversion to an array.  e.g. `Span<int> s = (int[])[1, 2, 3];`. 
+
+### BCL Apis
+
+BCL gave a status update on the work they have done.  Specifically: https://github.com/dotnet/runtime/pull/87945
+
+This (merged) PR adds the pattern of an immutable collection type `C<T>` having a factory method `public static C<T> C.Create<T>(ReadOnlySpan<T> values);` to create the collection.  A soon to be added attribute (`CollectionBuilderAttribute`) will be added that will allow the language/compiler to find this factory method from a given collection type.
+
+We do not plan any more BCL apis beyond this.  However, there are two key changes we will make in the compiler in terms of emit, to get even better codegen than what those above methods allow.  Specifically:
+
+1. while there is an `public static ImmutableArray<T> ImmutableArray.Create<T>(ReadOnlySpan<T> values);` method, the compiler will not call it.  Instead, it will sidestep it and use `ImmutableArray<T> ImmutableCollectionsMarshal.ToImmutableArray<T>(T[] values);` method.  This method passes ownership of the array *directly* to `ImmutableArray<T>` ensuring no overhead (no excess allocations *or* copying).
+
+2. while `List<T>` will already work with collection literals (due to it supporting collection-initializers) the compiler will emit new collection literal `List<T>'s` as:
+
+```c#
+List<int> list = [a, b, c]; // translates to:
+
+List<int> list = new List(capacity: 3);
+CollectionsMarshal.SetCount(list, 3);
+Span<int> __storage = CollectionsMarshal.AsSpan(list);
+__storage[0] = a;
+__storage[1] = b;
+__storage[2] = c;
+```
+
+Rule '1' will always apply for immutable arrays.  Rule '2' can apply depending on how complex it is for the compiler.  For example, if the code were `List<int> list] = [await a, await b, await c];` the compiler might choose to just emit this equivalently to `new List<int> { await a, await b, await c }` (i.e. equivalently to a collection initializer expression).
+
+### Breaking changes
+
+We would like to update the language so that `(A)[B]` is *always* a cast of a collection literal, not an indexing expression of a trivial identifier expression.  This would apply to `(A)[B]` (trivial identifeir), or `(A.B.C)[D]` (dotted identifier).  In both of these cases, if the user wanted indexing, it would be far more idiomatic to write this out as `A[B]` or `A.B.C[D]` (the parentheses are superfluous).
+
+THe compelling case to want to make this a cast of a collection literal is:
+
+```c#
+using CrazyArray = ImmutableArray<List<HashSet<int>>; // i.e. some complex collection type
```

Comment:
This might be the reason for the markdown source highlighter going wild
```suggestion
using CrazyArray = ImmutableArray<List<HashSet<int>>>; // i.e. some complex collection type
```